libs/layout/src/lib/services/media-query/mediaquery.service.ts

```diff
@@ -0,0 +1,103 @@
+import { Injectable, OnDestroy } from '@angular/core';
+import { WindowService } from '@ngx/utils';
+import { filter, map, Observable, ReplaySubject } from 'rxjs';
+
+/**
+ * A service that can be used to track media queries and their changes. It exposes a method
+ * to register media queries, which takes an array of tuples with the id of the media query
+ * and the query itself. The service will then emit the id of the media query that has
+ * changed when subscribed to the `getMatchingQuery$` method.
+ */
+@Injectable()
+export class NgxMediaQueryService implements OnDestroy {
+	/**
+	 * A map of media queries that are registered with the service.
+	 */
+	private readonly mediaQueryMap: Map<string, MediaQueryList> = new Map();
+
+	/*
+	 * A map of listeners that are registered with the service.
+	 * They are saved to be able to remove them when the service is destroyed.
+	 */
+	private readonly mediaQueryListenerMap: Map<
+		string,
+		(queryChangedEvent: MediaQueryListEvent) => void
+	> = new Map();
+
+	/**
+	 * A subject that emits the id of the media query that has changed.
+	 */
+	private readonly queryChangedSubject: ReplaySubject<string> = new ReplaySubject();
+
+	constructor(private readonly windowService: WindowService) {}
+
+	/**
+	 * Register a list of media queries that need to be tracked by the service.
+	 *
+	 * @param queries - A list of media queries that should be registered with the service.
+	 */
+	public registerMediaQueries(...queries: [id: string, query: string][]): void {
+		this.windowService.runInBrowser(({ browserWindow }) => {
+			for (let [id, query] of queries) {
```

Comment:
@IbenTesara remarked:
> I would add an early exit when a listener has already been registered on a certain `id`. Currently, the listener of a certain `id` will be overwritten without removing the previous one. This could lead to memory leaks.
> Example: setting two queries with the same id will result in the latter being destroyed in the onDestroy, while the first one will continue to exist and cause a memory leak.

Reply:
Good catch!